src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDocletWriter.java

```diff
@@ -1700,11 +1700,7 @@ private String redirectRelativeLinks(Element element, TextTree tt) {
         if (!shouldRedirectRelativeLinks(element)) {
             return text;
         }
-        String lower = Utils.toLowerCase(text);
-        if (lower.startsWith("mailto:")
-                || lower.startsWith("http:")
-                || lower.startsWith("https:")
-                || lower.startsWith("file:")) {
+        if (text.matches("^[^:/?#]+:.+$")) {
```

Comment:
Why not use `java.net.URI` API to determine if the link contains a scheme?

Reply:
I assume the link is in an HTML document and goes in an HTML document. Depending on where from `text` comes from and whereto it goes you might need first to decode it using URLDecoder, and then you would might need to reencode it before spitting it out... That's a lot of operations where things could go wrong, especially if the link contains a query string.

Reply:
That said a stricter regexp (unless I'm mistaken) could be: `^[a-zA-Z][a-zA-Z0-9+-.]*:.*$`
[ from RFC 2396:     scheme        = alpha *( alpha | digit | "+" | "-" | "." ) ]

Reply:
My only concerns were correctness and code reuse. Using an API doesn't require one to read through RFC.

Reply:
I'd argue for simply adding `ftp:` as an additional condition (unless there's other interesting URI schemes I'm not thinking of?). If a regex is to be used, I agree the regex should be much stricter (and defined in a constant, so that the `Pattern` doesn't need to be compiled on each invocation?).

Reply:
I would normally opt for a generic regexp-based solution such as proposed by @dfuch, but there is a security aspect to this as well (e.g. script invocation), so I'd go with the more conservative approach here to just add `ftp:` protocol to the list.